React EsBuild Starter Kit (#13)
metaplex-foundation/js-examples: getting-started-react-esbuild/README.md

# Getting Started with Metaplex and Esbuild

This example sets up a new React app with Metaplex using Esbuild.

This example has been generated using the following steps:

1. **Create a new project using NPM.**

>    Esbuild does not has any standard template for React which can be utilized for initializing the project like CRA or Vite. So, we will be creating a empty npm project and setup everything from scratch.

Reviewer: ```suggestion
   Esbuild does not have any standard template for React which can be utilized for initializing the project like CRA or Vite. So, we will be creating an empty npm project and setup everything from scratch.
```

   ```sh
   mkdir getting-started-react-esbuild
   cd getting-started-react-esbuild
   npm init -y
   ```

>    Once empty npm project is setup, we will setup a basic directory structure with mainly `public` and `src` folder and a config file (`build.js`) which will be used for compiling and bundling the react app with esbuild.

Reviewer: ```suggestion
   Once the empty npm project is ready, we will set up a basic directory structure with mainly `public` and `src` folders and a config file (`build.js`) which will be used for compiling and bundling the react app with esbuild.
```

   ```
   getting-started-react-esbuild/
   ├── public/
   │   └── index.html
   ├── src/
   │   ├── app.jsx
   │   ├── index.css
   │   └── index.js
   ├── build.js
   └── package.json
   ```

> 2. **Install the React, React DOM, EsBuild, Metaplex and the Solana SDKs.**

Reviewer: It's okay to focus on the important libraries when the list becomes too long.

```suggestion
2. **Install React, EsBuild and the Metaplex SDK.**
```

   ```sh
   npm install react react-dom esbuild @metaplex-foundation/js @solana/web3.js
   ```

3. **Install some polyfills.**

   <details>
     <summary>Why?</summary>
     Some dependencies of the Metaplex SDK are still relying on NPM packages that are not available in the browser. To make sure that the Metaplex SDK works in the browser, we need to install some polyfills.
   </details>

   ```sh
   npm install -D node-stdlib-browser
   ```

4. **Install Servor for Development**

    <details>
     <summary>Why?</summary>

>      Esbuild does not include any server that could be used to preview the esbuild built bundle in browser. So we will be using servor for that

Reviewer: ```suggestion
     Esbuild does not include any server that could be used to preview our app in the browser. So we will be using servor for that.
```

   </details>

   ```sh
   npm install -D servor
   ```

> 5. **Update your Esbuild build file and package.json**.

Reviewer: ```suggestion
5. **Update your build.js file**.
```

>    Add the following code to `build.js` file we created.

Reviewer: ```suggestion
   Add the following code to the `build.js` file we created earlier.
```

   <details>
     <summary>Why?</summary>

>      The following code will build and bundle your react app with EsBuild. It will also start a dev server when using in development

Reviewer: ```suggestion
     The following code will build and bundle your react app with EsBuild. It will also start a dev server when used in development.
```

   </details>

   ```js
   const esbuild = require('esbuild');
   const plugin = require('node-stdlib-browser/helpers/esbuild/plugin');
   const stdLibBrowser = require('node-stdlib-browser');
   const fs = require('fs');
   const path = require('path');
   const servor = require('servor');

   const outdirectory = 'public';

>    //clear out any old JS or CSS

Reviewer: ```suggestion
   // Clean previously built assets.
```

     });
     console.log('Development bundle built ✅');
     console.log('Running server from: http://localhost:8000');
     await servor({
       browser: true,
       root: outdirectory,
       port: 8000,
     });
   }

   async function prod() {
     console.log('Build started ⏳');
     await esbuild.build({
       entryPoints: ['src/index.js'],
       outdir: outdirectory,
       bundle: true,
       define: {
         'process.env.NODE_ENV': '"production"',
         global: 'global',
         process: 'process',
         Buffer: 'Buffer',
       },
       minify: true,
       inject: [require.resolve('node-stdlib-browser/helpers/esbuild/shim')],
       plugins: [plugin(stdLibBrowser)],
       loader: {
         '.js': 'jsx',
       },
     });
     console.log('Build completed ✅');
   }

   //defaults to build
   let config = '-build';
   if (process.argv.length > 2) {
     config = process.argv[2];
   }

   // Builds the bundle for dvelopment and runs a local web server
   // with livereload when -watch is set
   config === '-watch' && dev();

   // Builds optimized bundle for production
   config === '-build' && prod();
   ```

6. **Update your `package.json`.**

>    Add the following script in your `package.json` file.

Reviewer: ```suggestion
   Add the following scripts to your `package.json` file.
```
